bench/macro/lsm-tree-bench-wp8.hs

```diff
@@ -0,0 +1,404 @@
+            writeIORef keysRef $! IS.union
+                (IS.difference keys new)
+                (IS.fromList $ map fromIntegral inserts)
+
+        let (batch1, batch2) = toOperations lookups inserts
+        _ <- evaluate $ force (batch1, batch2)
+
+        return nextG
+
+    when opts.check $ do
+        duplicates <- readIORef duplicateRef
+        printf "True duplicates: %d\n" duplicates
+
+-------------------------------------------------------------------------------
+-- Batch generation
+-------------------------------------------------------------------------------
+
+{- | Implement generation of unbounded sequence of insert/delete operations
+
+matching UTxO style from spec: interleaved batches insert and lookup
+configurable batch sizes
+1 insert, 1 delete, 1 lookup per key.
+
+Current approach is probabilistic, but uses very little state.
+We could also make it exact, but then we'll need to carry some state around
+(at least the difference).
+
+-}
+generateBatch
+    :: Int       -- ^ initial size of the collection
+    -> Int       -- ^ batch size
+    -> SM.SMGen  -- ^ generator
+    -> Int       -- ^ batch number
+    -> (SM.SMGen, [Word64], [Word64])
+generateBatch initialSize batchSize g b = (nextG, lookups, inserts)
+  where
+    maxK :: Word64
+    maxK = fromIntegral $ initialSize + batchSize * b
+
+    lookups :: [Word64]
+    (!nextG, lookups) = mapAccumL (\g' _ -> swap (SM.bitmaskWithRejection64 maxK g')) g [1 .. batchSize]
+
+    inserts :: [Word64]
+    inserts = [ maxK .. maxK + fromIntegral batchSize - 1 ]
+
+-- | Generate operation inputs
+toOperations :: [Word64] -> [Word64] -> ([K], [(K, LSM.Update V B)])
+toOperations lookups inserts = (batch1, batch2)
+  where
+    batch1 :: [K]
+    batch1 = map makeKey lookups
+
+    batch2 :: [(K, LSM.Update V B)]
+    batch2 =
+        [ (k, LSM.Delete)
+        | k <- batch1
+        ] ++
+        [ (makeKey k, LSM.Insert theValue Nothing)
+        | k <- inserts
+        ]
+
+-------------------------------------------------------------------------------
+-- run
+-------------------------------------------------------------------------------
+
+doRun :: GlobalOpts -> RunOpts -> IO ()
+doRun gopts opts = do
+    time <- timed_ $ doRun' gopts opts
+    -- TODO: collect more statistic, save them in dry-run,
+    -- TODO: make the results human comprehensible.
+    printf "Proper run: %.03f sec\n" time
+
+doRun' :: GlobalOpts -> RunOpts -> IO ()
+doRun' gopts opts = do
+    let mountPoint :: FS.MountPoint
+        mountPoint = FS.MountPoint gopts.rootDir
+
+    let someFs :: FS.SomeHasFS IO
+        someFs = FS.SomeHasFS (FsIO.ioHasFS mountPoint)
+
+    name <- maybe (fail "invalid snapshot name") return $
+        LSM.mkSnapshotName "bench"
+
+    let initGen = SM.mkSMGen opts.seed
+
+    withSession someFs (FS.mkFsPath []) $ \session -> do
+        -- open snapshot
+        tbl <- LSM.open @IO @K @V @B session name
```

Comment:
It's very useful to be able to generate the initial LSM once and then reuse it, nice.

The snapshot API is only supposed to be implemented in WP15, but it makes sense to target it here (given that we only want to use the external API) and we should be able to come up with a hacky snapshot implementation as part of WP7 that isn't the real deal, but makes this benchmark work. :+1: